Add post-purchase onboardings to Recommendation flows

robertsreberski opened the pull request:
This PR **adds post-purchase onboardings to the Recommendations assistant** - it is being triggered for the most relevant purchase (out of supported ones - listed below) and overwrites the normal recommendations steps until it's completed.

<img width="1011" alt="image" src="https://user-images.githubusercontent.com/8419292/193134881-1225bcd4-dcf2-4e93-a371-26ffb63308a7.png">

**Supported purchases:**
- Jetpack Complete
- Jetpack Security
- Jetpack Backup
- Jetpack Anti-Spam
- Jetpack VideoPress
- Jetpack Search
- Jetpack Scan

<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Adds post-purchase onboarding mechanism
* Adds new assets together with new recommendations cards
* Adds new component (<PrimarySummary />) and modifies existing ones to support new onboardings

#### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
Project thread: pbNhbs-3dq-p2

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No.

#### Testing instructions:
* Remove all purchases connected to your test site (via Network Admin MC)
* Checkout this PR (e.g. `gh pr checkout 26484`)
* Go to WP Admin and `https://jetpack.com/redirect/?source=jetpack-plans&site=<your_test_site>`
* Purchase first of plans: Jetpack Scan (or Jetpack Complete that is the most complete onboarding - but will require you to remove the purchase in order to test next onboardings)
* You should be redirected to the recommendations page with a `Welcome to __!` card. E.g.:
<img width="1011" alt="image" src="https://user-images.githubusercontent.com/8419292/193134881-1225bcd4-dcf2-4e93-a371-26ffb63308a7.png">

* You should be able to go through the process, and all links should direct you to the valid pages on Jetpack Cloud or WP Admin dashboards
* The summary should include a summary of all features you've been presented. E.g.:

<img width="1003" alt="image" src="https://user-images.githubusercontent.com/8419292/193137841-f2069174-0309-4be4-8ff6-84a431706c29.png">

* Ensure all flows correspond to designs: p1HpG7-hqv-p2

* Best test plan: repeat the steps for all of 7 purchases in this order:

1. Jetpack Scan
2. Jetpack Search
3. Jetpack VideoPress
4. Jetpack AntiSpam
5. Jetpack Backup
6. Jetpack Security
7. Jetpack Complete

* Additionally: Try to purchase Jetpack Boost and ensure it doesn't break the recommendations flow and will not trigger any onboarding.

Every purchase made in this order should overwrite the previous onboarding. And if you already have e.g. Jetpack Complete, the purchase of the "lower" product should show you the onboarding process, but do not impact the summary on the last step - this should depend on the most relevant purchase.

robertsreberski committed:
Add initial onboarding mechanisms

robertsreberski committed:
changelog

robertsreberski committed:
Add components and content for JP Backup onboarding, improve  naming

robertsreberski committed:
Add onboarding progressValue

robertsreberski committed:
Clean summary content for onboardings

robertsreberski committed:
Update images

robertsreberski committed:
Add only skipped recommendations to skipped summary

robertsreberski committed:
Minor code improvements

robertsreberski committed:
Merge remote-tracking branch 'origin/trunk' into add/post-purchase-recommendations-onboardings

robertsreberski committed:
Add all new and changed assets

robertsreberski committed:
Add onboardings for all 7 purchases

robertsreberski committed:
Style primary summary

robertsreberski committed:
Adjust code to fix tests

robertsreberski committed:
Remove duplicate JSDoc @param

robertsreberski committed:
Merge branch 'trunk' into add/post-purchase-recommendations-onboardings

robertsreberski committed:
Include only new purchases in recommendations onboardings

robertsreberski commented:
I still plan to write unit tests supporting the new recommendations code to ensure it works as expected and does not break any existing functionalities. However, it should not prevent anyone from starting to review the PR!

robertsreberski committed:
Merge remote-tracking branch 'origin/trunk' into add/post-purchase-recommendations-onboardings

robertsreberski committed:
Revert eslint rules modification

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/recommendations/feature-utils.js:
```diff
@@ -340,6 +387,158 @@ export const getStepContent = stepSlug => {
+					'jetpack'
+				),
+				ctaText: __( 'Learn how to add videos to your site', 'jetpack' ),
+				ctaLink: getSiteAdminUrl( state ) + 'admin.php?page=jetpack-videopress',
```
This link doesn't seem to be working for me, this is what I get
![image](https://user-images.githubusercontent.com/65001528/193313954-79449112-619d-41d2-9740-ae2276b65bfa.png)
Not sure if that is some configuration on my end that is incorrect but I can't seem to fix it quickly at least

robertsreberski replied:
That's a good remark thanks! We introduced Jetpack VideoPress as a standalone plugin recently, and it indeed first needs to be installed before we can forward users to that page (I was having it activated by the mistake). I've introduced the code to first install the plugin (if necessary) before showing users this page.

I've also updated the current recommendations code (`/recommendations/videopress`) recommendation to support the standalone plugin as well.

CodeyGuyDylan replied:
@robertsreberski I am still having trouble on this plugin. I made sure I have all the most updated changes you made, and when I buy VideoPress it sends me straight here
![image](https://user-images.githubusercontent.com/65001528/193645397-7858c8dd-c002-42fd-bca1-23f45e62babe.png)
and when I click on the CTA I still get that same error screen
![image](https://user-images.githubusercontent.com/65001528/193645464-b356443f-96cd-4a83-abab-0677b766d0c5.png)

And then if I try the Add a Video button here
![image](https://user-images.githubusercontent.com/65001528/193647171-3aa851c1-4099-43c0-a206-af750dcc3e15.png)
I am getting an error in the console, I assume due to the videopress installation failing
![image](https://user-images.githubusercontent.com/65001528/193647239-63820740-02df-434f-8c0e-66e3c0ca39cb.png)

In addition to the error log, it also doesn't do anything on the screen when I press it

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/recommendations/feature-utils.js:
```diff
@@ -103,11 +108,52 @@ export const getSummaryResourceProps = resourceSlug => {
+			};
+		case 'videopress-activated':
+			return {
+				displayName: __( 'Ad-free, Customizable Video', 'jetpack' ),
```
What do we think about making all of these titles clickable to go back to the beginning of onboarding like we do with the feature and resource prompts?
![Screen Shot 2022-09-30 at 10 24 53 AM](https://user-images.githubusercontent.com/65001528/193314945-9505358c-70b7-4632-81dd-cacea1173e92.jpg)
Like this
![image](https://user-images.githubusercontent.com/65001528/193315121-596fa105-dbae-40ec-916a-b97ed905e7f0.png)

robertsreberski replied:
I'll ask the design team about that! I think it both has sense but also I understand that it can be undesired here (because it's basically a shortcut to the other page where you can get to know more).

robertsreberski committed:
Merge remote-tracking branch 'origin/trunk' into add/post-purchase-recommendations-onboardings

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/recommendations/constants.js:
```diff
@@ -14,8 +24,129 @@ export const RECOMMENDATION_WIZARD_STEP = {
 	ANTI_SPAM: 'anti-spam',
 	VIDEOPRESS: 'videopress',
 	BACKUP_PLAN: 'backup-plan',
+	WELCOME__BACKUP: 'welcome__backup',
```
Not a massive problem, just wondering, why are we using two underscores here between words instead of a single dash like the others?

robertsreberski replied:
I wanted to somehow distinguish the fact welcome steps are a bit different from the rest - I planned to underline that they belong to the special `welcome` category.

CodeyGuyDylan replied:
I see, kind of like how the `scss` classes are written. I think it makes sense, just looked a bit weird to me

robertsreberski replied:
Yeah, i think we will need to address things like this as we'll scale recommendations project

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/recommendations/constants.js:
```diff
@@ -14,8 +24,129 @@ export const RECOMMENDATION_WIZARD_STEP = {
+export function getOnboardingPriority( name ) {
+	const index = ONBOARDING_ORDER.indexOf( name );
+
+	return index > -1 ? index : 999;
```
An index of -1 would imply the name passed to the function does not exist in the onboarding, I'm not sure how much this matters but would returning `-1` instead of `999` make more sense? Since a non-existent onboarding value should have a very high priority, and would simplify the code a bit to `return index`

robertsreberski replied:
Yeah, I'm sorry for the confusion here, the sorting function prioritizes lesser numbers, so in that way `0` is the highest priority. I should have written in the documentation of the function `getOnboardingPriority` that `999` is the min priority.

Now I'm thinking if we don't want to have an exception thrown here if the onboarding is not found. In the end, if there will be only one onboarding in the array, and it won't exist, it will lead to errors further in the runtime anyway (not existing steps, etc.)

CodeyGuyDylan replied:
Ok that makes more sense now that you explain it, thanks

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/recommendations/feature-summary/primary.jsx:
```diff
@@ -0,0 +1,30 @@
+import { connect } from 'react-redux';
+import { getSummaryPrimaryProps } from '../feature-utils';
+
+const SummaryTextLink = ( { href, label } ) => (
```
I only see this function being used once, and not in any sort of loop. Was this to avoid too much nesting or do we plan to use in this in a loop or elsewhere in the future?

robertsreberski replied:
Yes, I wanted to avoid nesting in that way, it's a simple component only used once.

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/recommendations/feature-summary/primary.jsx:
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
```
You might want to consider making this a TypeScript file as I know we are shifting to using that instead of JavaScript. I know its a bit annoying to have mixed `.jsx` and `.tsx` files throughout the recommendations, but that already exists so we should probably lean toward new files being`.tsx`, what do you think?

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/recommendations/feature-summary/style.scss:
```diff
@@ -14,6 +14,33 @@
+		$spacing: 0.5rem;
+
+		padding-top: $spacing;
+		padding-bottom: $spacing;
```
`padding: $spacing 0` is a bit more concise in this case, and since the `$spacing` variable would only be used once you could remove it entirely and just use `padding: 0.5rem 0`. This is just what I would do and feel free to let me know if you had a good reason for doing it this way 😄

robertsreberski replied:
The `is-primary` class is a modifying class. I haven't used the shorthand for padding as I wanted to restrain from overwriting horizontal padding when unnecessary - to make sure it will still inherit values from the parent if we'll e.g. decide to change overall styling a bit.

CodeyGuyDylan replied:
Ah ok yes, this makes sense, thanks!

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/recommendations/feature-summary/primary.jsx:
```diff
@@ -0,0 +1,30 @@
+		{ label }
+		<span class="jp-summary-text-link__icon dashicons dashicons-arrow-right-alt2"></span>
+	</a>
+);
```
Another comment on this, would it be better for these links to look the same as the other buttons on the summary page? Like this:
![image](https://user-images.githubusercontent.com/65001528/193331640-6a228cb3-3f3d-4516-9634-f32f46621251.png)
instead of this:
![image](https://user-images.githubusercontent.com/65001528/193331687-a5a3b157-a24a-4360-868b-0f8747ca9da0.png)

Or were these designed to be vastly different as to separate them visually from the rest?

robertsreberski replied:
They have been designed that way on purpose, as I understand to appear less actionable than the recommendations CTAs. I will ask the design team for confirmation.

CodeyGuyDylan replied:
I figured that would be a possibility, was just making sure 😄

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/state/recommendations/reducer.js:
```diff
@@ -332,16 +405,40 @@ export const stepToRoute = {
 
 export const isStepViewed = ( state, featureSlug ) => {
-	const recommendationsData = get( state.jetpack, [ 'recommendations', 'data' ] );
+	const recommendationsData = getRecommendationsData( state );
```
Nice improvement here 🏆

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/state/recommendations/reducer.js:
```diff
@@ -467,38 +566,112 @@ const isStepEligibleToShow = ( state, step ) => {
+		case 'videopress-activated':
+		case 'search-activated':
+		case 'server-credentials':
+			return true; // Can we check if credentials were added?
```
I am not sure about that, you may want to ask one of the relevant slack channels about this to see if there is. Someone will know 😄

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/state/recommendations/reducer.js:
```diff
@@ -467,38 +566,112 @@ const isStepEligibleToShow = ( state, step ) => {
+	const stepToNextStep = get( stepToNextStepByPath, active ? `onboarding.${ active }` : 'default' );
+
+	if ( ! stepToNextStep ) {
+		throw 'Next eligible steps: No path found!';
```
I wonder if we could send them to the summary page in this scenario instead of throwing an error. Would probably be a better user experience if something were to happen that triggered this

CodeyGuyDylan reviewed (commented):
It seems a bit weird to show multiple onboarding steps here when you have complete
![image](https://user-images.githubusercontent.com/65001528/193327180-d2c405b7-87b8-4889-b0d5-38a43a07d159.png)
But when you buy several single plans (ex. backup and then anti-spam) it only shows the spam onboarding in the summary. Would it be possible/preferable to show both backup and spam at the same time if you have both?

My last comment is that next time we have a change this big, it might be better to add the framework for the new functionality in one PR (i.e. adding the ability to add onboarding steps) and separate each new onboarding plan as a separate PR. I think that would make it a bit easier to review and look at in the future instead of one giant PR 😄

CodeyGuyDylan reviewed (changes requested):
Forgot to officially request changes on the actual review 😆

robertsreberski commented:
> But when you buy several single plans (ex. backup and then anti-spam) it only shows the spam onboarding in the summary. Would it be possible/preferable to show both backup and spam at the same time if you have both?

Thanks for pointing that out, I'll confirm it with design, but it indeed makes sense. 

> My last comment is that next time we have a change this big, it might be better to add the framework for the new functionality in one PR (i.e. adding the ability to add onboarding steps) and separate each new onboarding plan as a separate PR. I think that would make it a bit easier to review and look at in the future instead of one giant PR 😄

Yes, sorry for that. I'm actually torn between approaches to situations like this. Because on one hand, that way it's not possible to functionally test the mechanisms, but on the other - we end up with lots of code to view at once. Probably it would be great to leave a working MVP (based on single onboarding) as one PR, and add more onboardings as a separate one.

As a side note, I tried to still structure commits in an atomic way. You should be able to see MVP code by only selecting [these 9 commits](https://github.com/Automattic/jetpack/pull/26484/files/cbbda7730877f0908d95808f6c5447dcc0426fde). And then the further commits focus on adding the next onboardings.

I agree it's not a perfect solution and will improve the PR structure in the next projects.

CodeyGuyDylan commented:
> Yes, sorry for that.

No worries, I totally see the reasoning for doing it this way

> Probably it would be great to leave a working MVP (based on single onboarding) as one PR, and add more onboardings as a separate one.

I think that would be a great solution

robertsreberski committed:
Adjust recommendations to new standalone jetpack-videopress plugin

robertsreberski committed:
Install necessary plugins (if necessary) before directing to plugin page

robertsreberski committed:
Support edge cases in onboarding and step getters

robertsreberski committed:
Add typing to <PrimarySummary />

robertsreberski committed:
Add Tracks tracking to <PrimarySummary />

robertsreberski committed:
Ensure to display onboarding steps only if they belong to active onboarding

robertsreberski committed:
Remove TODO comment about Tracks

robertsreberski committed:
Display non-overlapping onboardings summaries

robertsreberski committed:
Move helper functions to separate file

robertsreberski committed:
Display server credentials step only when they haven't been provided yet

robertsreberski committed:
Add anchor to step name in primary summary

robertsreberski committed:
Merge branch 'trunk' into add/post-purchase-recommendations-onboardings

robertsreberski commented:
Hey @CodeyGuyDylan! Thanks again for the review. I've addressed all your feedback and I'll be adding unit tests to these changes, as a separate PR based on this one - not to further complicate the review process.

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/recommendations/feature-summary/primary.tsx:
```diff
@@ -0,0 +1,97 @@
+import Button from 'components/button';
+import analytics from 'lib/analytics';
+import React, { useCallback } from 'react';
```
I don't actually know which way is better but I like to avoid importing React in entirety since it is no longer required by React, I'd personally do
`import type { MouseEventHandler } from 'react';`
instead, but this is definitely not a blocker in any way

robertsreberski replied:
Thanks for that! I think it will make the code cleaner 👍

CodeyGuyDylan replied:
Not a big thing, but I think you may have forgotten to change this 😄

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/state/recommendations/reducer.js:
```diff
@@ -6,7 +6,7 @@ import {
 	PLAN_JETPACK_BACKUP_T1_YEARLY,
 } from 'lib/plans/constants';
-import { assign, difference, get, isArray, isEmpty, mergeWith, union } from 'lodash';
+import { assign, difference, get, isArray, isEmpty, mergeWith, union, intersection } from 'lodash';
```
I think we are moving away from lodash in Calypso: p4TIVU-9Bf-p2 . I can't find if we are doing this in the plugin, but I thought I'd note it here just in case

CodeyGuyDylan reviewed (changes requested):
I am encountering a couple of errors that I am not entirely sure why are happening. I commented the first one on the previous comment about them, it has to do with the videopress plugin again. 

I noticed some weird behavior when going through onboarding when I had more than 1 standalone plan (in this case it was JP Scan and JP Anti-Spam. I already had Scan, and then I purchased JP Anti-Spam. When I got directed to onboarding, I saw the JP Scan onboarding again and never saw the scan onboarding.

![image](https://user-images.githubusercontent.com/65001528/193650555-69f7dd17-8df5-46dc-89fe-6fed066df565.png)

and if i press next

![image](https://user-images.githubusercontent.com/65001528/193650610-b1237661-4155-4ce1-bfee-8e3837c441e4.png)

Could be something wrong with the way onboarding is selected when more than one single plugin plan is active

robertsreberski committed:
Change cta link on videopress step

robertsreberski committed:
Ensure that viewed onboardings are still eligible ones

robertsreberski committed:
Redirect to site-type question on the next visit to recommendations (after finishing onboarding)

robertsreberski committed:
Merge remote-tracking branch 'origin/trunk' into add/post-purchase-recommendations-onboardings

robertsreberski committed:
Disable antispam and videopress steps if plugin is not installed

robertsreberski committed:
Remove redundant check to show steps

robertsreberski commented:
@CodeyGuyDylan thanks again for taking the time to provide me feedback on the code! I've addressed your comments and made the needed improvements. I hope everything works as expected now.

robertsreberski committed:
Fix unit tests

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/recommendations/onboarding-utils.ts:
```diff
@@ -0,0 +1,47 @@
+	const index = ONBOARDING_ORDER.indexOf( name );
+
+	if ( -1 === index ) {
+		throw `The onboarding "${ name }" is not included in the ONBOARDING_ORDER`;
```
I bought Boost just to see what would happen, and this error got thrown.
![image](https://user-images.githubusercontent.com/65001528/194897590-0b1fd08a-36ef-40f1-8f87-6010b145e630.png)

Seems like even though we don't have an onboarding for Boost, this function still got called and was not handled to where it avoided a fatal error. Perhaps we can return something here like "unknown" or `null` to tell whatever calls it to just send to the summary page instead? Or perhaps not even redirect to the assistant when boost is purchased.

CodeyGuyDylan replied:
CRM has a different site to purchase it on and Social can't be bought yet, but I assume they would react the same way if they were active/purchased through the same workflow

CodeyGuyDylan replied:
Additional info, I went and bought Jetpack Security after this and got more errors, probably due to still searching for Boost, because I went and removed Boost from my site and they went away
![image](https://user-images.githubusercontent.com/65001528/194898524-a78fbffa-eacf-4c3f-9cde-5263a0fcf1fa.png)

robertsreberski replied:
Thanks for spotting that one! It's indeed a very crucial one, I'm addressing it.

CodeyGuyDylan commented on projects/plugins/jetpack/_inc/client/recommendations/feature-utils.js:
```diff
@@ -170,7 +221,9 @@ export const mapDispatchToProps = ( dispatch, featureSlug ) => {
 			return {
 				activateFeature: () => {
-					return dispatch( updateSettings( { videopress: true } ) );
+					return restApi.installPlugin( 'videopress', 'recommendations' ).then( () => {
```
Was this working? Or are we circumventing that by getting rid of the CTA if videopress is not installed [here](https://github.com/Automattic/jetpack/pull/26484/files#diff-27a0ea9326bc6bff96c64d327b5f9bb7b25b819511b6e35033cc4522f2ff7ac8R149)

robertsreberski replied:
It stopped working after the introduction of a standalone JP VideoPress plugin. So I've updated it to the recent situation.

CodeyGuyDylan replied:
Ok that is what I thought, thanks!

CodeyGuyDylan reviewed (changes requested):
I found one more bug because I happened to purchase boost as well ( I had not done that previously), other than this, everything worked and looks great!

robertsreberski committed:
Import react types directly

robertsreberski committed:
Ensure that null is not returned as part of eligible onboardings

CodeyGuyDylan reviewed (dismissed):
Fixes look good, LGTM!

robertsreberski committed:
Fix missing backup card and change label for anti-spam cta

robertsreberski committed:
Add unit tests for recommendation onboardings (#26676)

* Add tests for onboardings to <Summary /> component

* Translate header text in onboarding summaries section

* Add tests to recommendations selectors and onboarding-utils

robertsreberski committed:
Implement cosmetic changes to recommendations (#26711)

* Consider onboarding steps in unviewed recommendations count

* Load recommendations data (incl. onboarding) in all other tabs to display valid count

* Hide recommendations count when the recommendations tab is active

* Add progress text to onboarding progress

robertsreberski committed:
Merge branch 'trunk' into add/post-purchase-recommendations-onboardings

coder-karen committed:
Merge branch 'trunk' into add/post-purchase-recommendations-onboardings

coder-karen committed:
Merge branch 'trunk' into add/post-purchase-recommendations-onboardings

coder-karen committed:
Merge branch 'trunk' into add/post-purchase-recommendations-onboardings

coder-karen reviewed (commented):
When adding multiple plans (one by one) to test this out, I notice that if I enable one of the recommendations such as Site Accelerator in one flow, the Site Accelerator step is still counted in the number of steps for any new recommendation flows (but that card is skipped). So I may see myself on step 2 of 4 (Downtime monitoring for example), then the next step is all of a sudden step 4. I'm left wondering what step 3 was :) I wonder if the number of steps can be modified to remove recommendations that have already been added?

Separately, I removed and re-added Search, and didn't get the Search onboarding flow the second time but instead a Protect recommendation. If I've already seen the recommendation flow earlier, is that expected the second time around?

One last thing - I noticed in the Complete onboarding recommendations flow that the Next button changed after the first step to have a border around it. Is that expected?

Step 1:
<img width="237" alt="Screen Shot 2022-10-14 at 10 59 25 am" src="https://user-images.githubusercontent.com/16754605/195823201-f18ad459-592d-4b6d-8928-ae155d08123d.png">

Step 2: 
<img width="301" alt="Screen Shot 2022-10-14 at 10 59 32 am" src="https://user-images.githubusercontent.com/16754605/195823268-a7f69296-a4ee-402a-94bd-423db08f3124.png">

coder-karen committed:
Merge branch 'trunk' into add/post-purchase-recommendations-onboardings

robertsreberski committed:
Merge branch 'trunk' into add/post-purchase-recommendations-onboardings

robertsreberski committed:
Sync viewed onboardings even when no onboarding active

robertsreberski committed:
Fix flag state

robertsreberski committed:
Add info about total steps to onboarding (to calculate progress)

robertsreberski committed:
Replace arrays in onboarding data instead of merging

robertsreberski committed:
Disable focus state for navigation links

robertsreberski committed:
Merge branch 'trunk' into add/post-purchase-recommendations-onboardings

robertsreberski commented:
Thanks for checking out the PR @coder-karen! I've addressed all the unpolishments you've spotted.

robertsreberski committed:
Fix unit tests

coder-karen committed:
Merge branch 'trunk' into add/post-purchase-recommendations-onboardings

coder-karen committed:
Merge branch 'trunk' into add/post-purchase-recommendations-onboardings

coder-karen reviewed (commented):
Thanks @robertsreberski ! I had another look now - the next buttons look good (when adding the Complete plan) and the steps work well!

I tried removing and re-adding the Search plan again, this time I ended up with a recommendation to add credentials (that was a step I hadn't completed on the only other existing plan at the time - Scan). It looks like plan recommendations don't necessarily re-start when re-adding a plan, and they perhaps default to recommendations for other plans? Then again, I'm not sure how common it would be for a customer to actively remove and re-add a plan so this could be more of a nitpick. Still, would this be a currently expected behaviour to have recommendations correctly restart for re-added plans?

robertsreberski committed:
Start new onboarding even when the previous is currently active

robertsreberski committed:
Merge branch 'add/post-purchase-recommendations-onboardings' of github.com:Automattic/jetpack into add/post-purchase-recommendations-onboardings

robertsreberski commented:
@coder-karen thanks for taking a second look at the PR!

> It looks like plan recommendations don't necessarily re-start when re-adding a plan, and they perhaps default to recommendations for other plans?

Thanks for explaining the steps you've taken. What was happening is that the code did not support activating new onboarding when there was any other one already active. Indeed, we want to show to the user the most recent onboarding, without waiting for them to finish the previous one - I've provided the change to allow for that now. It wasn't an issue, just an overseen case - thanks for spotting that!

Note that, after removing the plan, you need to access the Jetpack dashboard at least once before purchasing it again, in order to sync viewed onboardings. I don't think it is much of an issue, as if the user re-purchases the package so quickly, they probably do not need to see the onboarding again.

coder-karen committed:
Merge branch 'trunk' into add/post-purchase-recommendations-onboardings

coder-karen reviewed (approved):
@robertsreberski thanks for clarifying and the additional change!
I tried re-adding Anti-Spam and I then saw the Anti-Spam recommendation which is perfect.
Looks good from my end 👍
